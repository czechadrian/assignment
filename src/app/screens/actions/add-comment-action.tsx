import { ActionType, createAction } from "typesafe-actions";
import { TComment } from "../../reducers/comments";

export const addCommentAction = createAction("comment/ADD_NEW_COMMENT")<
  TComment
>();

export type TAddCommentAction = ActionType<typeof addCommentAction>;
