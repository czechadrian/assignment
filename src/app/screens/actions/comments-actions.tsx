import { ActionType, createAction } from "typesafe-actions";
import { ThunkAction } from "redux-thunk";
import { TRootState } from "../../reducers";
import { getComments } from "../../../api-wrapper/get-comments";
import { TComments } from "../../reducers/comments";

export const getCommentsInitAction = createAction("data/FETCH_INIT")();
export const getCommentsSuccessAction = createAction("data/FETCH_SUCCESS")<
  Pick<TComments, "comments">
>();
export const getCommentsFailureAction = createAction("data/FETCH_FAILURE")();

export type TGetCommentsInitAction = ActionType<typeof getCommentsInitAction>;
export type TGetCommentsSuccessAction = ActionType<
  typeof getCommentsSuccessAction
>;
export type TGetCommentsFailureAction = ActionType<
  typeof getCommentsFailureAction
>;

export type TGetCommentsActions =
  | TGetCommentsSuccessAction
  | TGetCommentsFailureAction
  | TGetCommentsInitAction;

export type TFetchCommentsAction = () => ThunkAction<
  void,
  TRootState,
  null,
  TGetCommentsActions
>;
export const fetchComments: TFetchCommentsAction = () => (dispatch) => {
  dispatch(getCommentsInitAction());
  return getComments()
    .then((payload) => {
      return dispatch(getCommentsSuccessAction({ comments: payload }));
    })
    .catch(() => {
      dispatch(getCommentsFailureAction());
    });
};
