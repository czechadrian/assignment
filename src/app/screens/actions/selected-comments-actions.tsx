import { ActionType, createAction } from "typesafe-actions";
import { TComment } from "../../reducers/comments";

export const selectCommentAction = createAction(
  "homepage/CLICK_SELECT_COMMENT"
)<Pick<TComment, "id">>();
export const removeSelectedCommentAction = createAction(
  "selected-comments/REMOVE_SELECTED_COMMENT"
)<Pick<TComment, "id">>();

export type TSelectCommentAction = ActionType<typeof selectCommentAction>;

export type TRemoveSelectedCommentAction = ActionType<
  typeof removeSelectedCommentAction
>;

export type TSelectedCommentActions =
  | TSelectCommentAction
  | TRemoveSelectedCommentAction;
