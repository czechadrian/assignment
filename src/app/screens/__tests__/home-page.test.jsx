import React from "react";
import "jest-styled-components";
import renderer from "react-test-renderer";
import HomePage from "../home-page";
import { useSelector, useDispatch } from "react-redux";
import { selectCommentAction } from "../actions/selected-comments-actions";
import { mount } from "enzyme";
import { useHistory } from "react-router-dom";
const { payload } = require("../../../api-wrapper/test-helper/comments.data");
jest.mock("../actions/selected-comments-actions", () => {
  return {
    selectCommentAction: jest.fn(),
  };
});
jest.mock("react-redux", () => {
  return {
    useSelector: jest.fn(),
    useDispatch: jest.fn(),
  };
});
jest.mock("react-router-dom", () => {
  return {
    useHistory: jest.fn(),
  };
});

describe("HomePage", () => {
  const historyMock = jest.fn();
  const dispatchMock = jest.fn();
  beforeEach(() => {
    useDispatch.mockReturnValue(dispatchMock);
    useHistory.mockReturnValue(historyMock);
    useSelector.mockImplementation((selectorFn) =>
      selectorFn({ data: payload })
    );

    selectCommentAction.mockReturnValue(() => jest.fn());

    jest.clearAllMocks();
  });
  test("should match snapshot HomePage component", () => {
    const tree = renderer.create(<HomePage />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("should test styles HomePage component with empty comments collection", () => {
    useSelector.mockImplementation((selectorFn) =>
      selectorFn({ data: { comments: [] } })
    );
    const tree = mount(<HomePage />);

    expect(tree.contains("Loading....")).toBe(true);
  });

  test("should test styles HomePage component with non-empty comments collection", () => {
    const tree = mount(<HomePage />);

    expect(tree.find("main header div").exists()).toBe(true);

    const button1 = tree.find("main header div button").at(0);
    const button2 = tree.find("main header div button").at(1);
    const button3 = tree.find("main header div button").at(2);

    expect(button1.exists()).toBe(true);
    expect(button2.exists()).toBe(true);
    expect(button3.exists()).toBe(true);
  });
});
