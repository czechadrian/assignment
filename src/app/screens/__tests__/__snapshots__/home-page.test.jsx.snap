// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage should match snapshot HomePage component 1`] = `
.c0 {
  width: 80%;
  margin: 0 10% 5%;
  border-bottom: 1px solid lightgray;
}

.c1:last-child {
  margin-bottom: 2.5%;
}

.c4 {
  align: right;
}

.c3 {
  aria-label: "simple table";
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0 10% 0;
}

<main>
  <header
    className="c0"
  >
    <div
      className="MuiBottomNavigation-root c1"
    >
      <button
        className="MuiButtonBase-root MuiBottomNavigationAction-root Mui-selected"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiBottomNavigationAction-wrapper"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
          <span
            className="MuiBottomNavigationAction-label Mui-selected"
          >
            Homepage
          </span>
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiBottomNavigationAction-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiBottomNavigationAction-wrapper"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2h-8l-2-2z"
            />
          </svg>
          <span
            className="MuiBottomNavigationAction-label"
          >
            Selected comments
          </span>
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiBottomNavigationAction-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiBottomNavigationAction-wrapper"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 2H4c-1.1 0-1.99.9-1.99 2L2 22l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm-2 12H6v-2h12v2zm0-3H6V9h12v2zm0-3H6V6h12v2z"
            />
          </svg>
          <span
            className="MuiBottomNavigationAction-label"
          >
            Add comment
          </span>
        </span>
      </button>
    </div>
  </header>
  <main
    className="c2"
  >
    <div
      className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <table
        className="MuiTable-root c3"
        role={null}
      >
        <thead
          className="MuiTableHead-root"
          role={null}
        >
          <tr
            className="MuiTableRow-root MuiTableRow-head"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head c4"
              scope="col"
            >
              Name
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head c4"
              scope="col"
            >
              E-mail
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head c4"
              scope="col"
            >
              First 20 characters
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head c4"
              scope="col"
            >
              Add/Remove to Selected
            </th>
          </tr>
        </thead>
        <tbody
          className="MuiTableBody-root"
          role={null}
        >
          <tr
            className="MuiTableRow-root"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
              role="cell"
              scope="row"
            >
              id labore ex et quam laborum
            </th>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body c4"
            >
              Eliseo@gardner.biz
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body c4"
            >
              laudantium enim quas
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body c4"
            >
              <button
                className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
                disabled={true}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={-1}
                type="button"
              >
                <span
                  className="MuiButton-label"
                >
                  Added
                </span>
              </button>
            </td>
          </tr>
          <tr
            className="MuiTableRow-root"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
              role="cell"
              scope="row"
            >
              quo vero reiciendis velit similique earum
            </th>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body c4"
            >
              Jayne_Kuhic@sydney.com
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body c4"
            >
              est natus enim nihil
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body c4"
            >
              <button
                className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiButton-label"
                >
                  Add
                </span>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </main>
</main>
`;
